src/storage/data_table.cpp

```diff
@@ -246,6 +274,189 @@ bool DataTable::Delete(transaction::TransactionContext *const txn, const TupleSl
+    outfile.write(ALIGNMENT, padded_flatbuf_size - flatbuf_size);
+  }
+  outfile.flush();
+}
+
+void DataTable::WriteDictionaryMessage(std::ofstream &outfile, int64_t dictionary_id,
+                                       const ArrowVarlenColumn &varlen_col,
+                                       flatbuffers::FlatBufferBuilder *flatbuf_builder) const {
+  std::vector<flatbuf::FieldNode> field_nodes;
+  std::vector<flatbuf::Buffer> buffers;
+  uint32_t num_elements = varlen_col.OffsetsLength() - 1;
+  size_t buffer_offset = 0;
+  field_nodes.emplace_back(num_elements, 0);
+  char dump_buffer[common::Constants::BLOCK_SIZE];
+
+  // Fake validity buffer
+  buffers.emplace_back(buffer_offset, 0);
+
+  MemoryBlockToBuffer(reinterpret_cast<const char *>(varlen_col.Offsets()), dump_buffer, &buffer_offset,
+                      varlen_col.OffsetsLength() * sizeof(uint64_t), &buffers);
+
+  MemoryBlockToBuffer(reinterpret_cast<const char *>(varlen_col.Values()), dump_buffer, &buffer_offset,
+                      varlen_col.ValuesLength(), &buffers);
+
+  auto record_batch =
+      flatbuf::CreateRecordBatch(*flatbuf_builder, num_elements, flatbuf_builder->CreateVectorOfStructs(field_nodes),
+                                 flatbuf_builder->CreateVectorOfStructs(buffers));
+  auto dictionary_batch = flatbuf::CreateDictionaryBatch(*flatbuf_builder, dictionary_id, record_batch);
+  auto aligned_offset = AlignedSize<size_t>(ARROW_ALIGNMENT, buffer_offset);
+  auto message =
+      flatbuf::CreateMessage(*flatbuf_builder, flatbuf::MetadataVersion_V4, flatbuf::MessageHeader_DictionaryBatch,
+                             dictionary_batch.Union(), aligned_offset);
+  flatbuf_builder->Finish(message);
+  int32_t flatbuf_size = flatbuf_builder->GetSize();
+  auto padded_flatbuf_size = AlignedSize<int32_t>(ARROW_ALIGNMENT, flatbuf_size);
+  outfile.write(reinterpret_cast<const char *>(&FLATBUF_CONTINUZATION), sizeof(int32_t));
+  outfile.write(reinterpret_cast<const char *>(&padded_flatbuf_size), sizeof(int32_t));
+  outfile.write(reinterpret_cast<const char *>(flatbuf_builder->GetBufferPointer()), flatbuf_size);
+  if (padded_flatbuf_size != flatbuf_size) {
+    outfile.write(ALIGNMENT, padded_flatbuf_size - flatbuf_size);
+  }
+  outfile.write(dump_buffer, aligned_offset);
+  outfile.flush();
+}
+
+void DataTable::ExportTable(const std::string &file_name) const {
+  flatbuffers::FlatBufferBuilder flatbuf_builder;
+  std::ofstream outfile(file_name, std::ios_base::out | std::ios_base::trunc | std::ios_base::binary);
+  std::unordered_map<col_id_t, int64_t> dictionary_ids;
+  WriteSchemaMessage(outfile, &dictionary_ids, &flatbuf_builder);
+
+  const BlockLayout &layout = accessor_.GetBlockLayout();
+  auto column_ids = layout.AllColumns();
+  common::SpinLatch::ScopedSpinLatch guard(&blocks_latch_);
+  // We may choose to scan twice instead of using a buffer?
+  auto dump_buffer = new char[MAX_DUMP_BUFFER_BLOCK * common::Constants::BLOCK_SIZE];
```

Comment:
The purpose of this `dump_buffer` use in these methods is not clear. Why are we copying into this, and then writing it out to the file? Copying seem's like an expensive step that we can just avoid by writing to the file directly. 

Your comment seems to indicate that you can scan once for all the metadata you need, write the metadata, and then write out the data directly to file without having to use this buffer? I would run perf and confirm it, but my intuition tells me that using this `dump_buffer` is expensive. It's also like 10MB alone which seems like a large amount of stack space to allocate.

Reply:
Changed to two-pass writing.